Update Routing to 6.0 (#23932)
dotnet/AspNetCore.Docs: aspnetcore/fundamentals/routing.md

## Routing basics

The following code shows a basic example of routing:

:::code language="csharp" source="routing/samples/6.0/RoutingSample/Program.cs" highlight="4":::

The preceding example includes a single endpoint using the [MapGet](xref:Microsoft.AspNetCore.Builder.EndpointRouteBuilderExtensions.MapGet%2A) method:

* When an HTTP `GET` request is sent to the root URL `/`:
  * The request delegate executes.
  * `Hello World!` is written to the HTTP response.
* If the request method is not `GET` or the root URL is not `/`, no route matches and an HTTP 404 is returned.

Routing uses a pair of middleware, registered by <xref:Microsoft.AspNetCore.Builder.EndpointRoutingApplicationBuilderExtensions.UseRouting%2A> and <xref:Microsoft.AspNetCore.Builder.EndpointRoutingApplicationBuilderExtensions.UseEndpoints%2A>:

* `UseRouting` adds route matching to the middleware pipeline. This middleware looks at the set of endpoints defined in the app, and selects the [best match](#urlm) based on the request.
* `UseEndpoints` adds endpoint execution to the middleware pipeline. It runs the delegate associated with the selected endpoint.

> Apps that use [Minimal APIs](xref:fundamentals/minimal-apis) typically don't need to call `UseRouting` or `UseEndpoints`. <xref:Microsoft.AspNetCore.Builder.WebApplicationBuilder> configures a middleware pipeline that wraps middleware added in *Program.cs* with `UseRouting` and `UseEndpoints`. However, apps that use Minimal APIs can change the order in which `UseRouting` and `UseEndpoints` run by calling these methods explicitly. For example, fhe following code makes an explicit call to `UseRouting`:

Reviewer: Shouldn't this change this to ASP.NET Core apps? That applies to all apps. I don't want to give the impression you need to call UseRouting or UseEndpoints.  in other apps.

Author: I dunno... We're in the context of ASP.NET Core here, so I thought it might be redundant to state that here. I can add it if you think it helps, of course. Just make the suggestion and I'll commit it with the other one.

Reviewer: I think someone could infer minimal api's were different given that's what you're calling out. How about

```suggestion
ASP.NET Core apps typically don't need to call `UseRouting` or `UseEndpoints`. <xref:Microsoft.AspNetCore.Builder.WebApplicationBuilder> configures a middleware pipeline that wraps middleware added in *Program.cs* with `UseRouting` and `UseEndpoints`. However, apps can change the order in which `UseRouting` and `UseEndpoints` run by calling these methods explicitly. For example, fhe following code makes an explicit call to `UseRouting`:
```

Author: Ok, I think I understand what you're going for now. Can we just say "Apps typically don't need..."? That doesn't call out Minimal APIs as being special, and also doesn't repeat ASP.NET Core that's already a given imo.

The following example shows routing with a more sophisticated route template:

:::code language="csharp" source="routing/samples/6.0/RoutingSample/Snippets/Program.cs" id="snippet_RouteTemplate":::

The string `/hello/{name:alpha}` is a **route template**. A route template is used to configure how the endpoint is matched. In this case, the template matches:

* A URL like `/hello/Docs`
* Any URL path that begins with `/hello/` followed by a sequence of alphabetic characters. `:alpha` applies a route constraint that matches only alphabetic characters. [Route constraints](#route-constraints) are explained later in this article.

The second segment of the URL path, `{name:alpha}`:

* Is bound to the `name` parameter.
* Is captured and stored in [HttpRequest.RouteValues](xref:Microsoft.AspNetCore.Http.HttpRequest.RouteValues%2A).

The following example shows routing with [health checks](xref:host-and-deploy/health-checks) and authorization:

:::code language="csharp" source="routing/samples/6.0/RoutingSample/Snippets/Program.cs" id="snippet_HealthChecksAuthz":::

> [!INCLUDE[request localized comments](~/includes/code-comments-loc.md)]

Reviewer: ```suggestion
```

> 

Author: ```suggestion
```

The following code shows how to retrieve and inspect the endpoint matching the current request:

:::code language="csharp" source="routing/samples/6.0/RoutingSample/Snippets/Program.cs" id="snippet_InspectEndpointMiddleware":::

The endpoint, if selected, can be retrieved from the `HttpContext`. Its properties can be inspected. Endpoint objects are immutable and cannot be modified after creation. The most common type of endpoint is a <xref:Microsoft.AspNetCore.Routing.RouteEndpoint>. `RouteEndpoint` includes information that allows it to be selected by the routing system.

In the preceding code, [app.Use](xref:Microsoft.AspNetCore.Builder.UseExtensions.Use%2A) configures an inline [middleware](xref:fundamentals/middleware/index).

<a name="mt"></a>

The following code shows that, depending on where `app.Use` is called in the pipeline, there may not be an endpoint:

:::code language="csharp" source="routing/samples/6.0/RoutingSample/Snippets/Program.cs" id="snippet_CurrentEndpointMiddlewareOrder":::

> This preceding sample adds `Console.WriteLine` statements that display whether or not an endpoint has been selected. For clarity, the sample assigns a display name to the provided `/` endpoint. Apps that use [Minimal APIs](xref:fundamentals/minimal-apis) typically don't need to call `UseRouting` or `UseEndpoints`. The preceding sample code includes these calls to control exactly when the `UseRouting` and `UseEndpoints` middleware runs within the pipeline.

Author: ```suggestion
The preceding sample adds `Console.WriteLine` statements that display whether or not an endpoint has been selected. For clarity, the sample assigns a display name to the provided `/` endpoint.

The preceding sample also includes calls to `UseRouting` and `UseEndpoints` to control exactly when these middleware run within the pipeline.
```

* A segment with literal text is considered more specific than a parameter segment.
* A parameter segment with a constraint is considered more specific than one without.
* A complex segment is considered as specific as a parameter segment with a constraint.
* Catch-all parameters are the least specific. See **catch-all** in the [Route templates](#rtr) section for important information on catch-all routes.

> See the [source code on GitHub](https://github.com/dotnet/aspnetcore/blob/main/src/Http/Routing/src/Template/RoutePrecedence.cs#L189) for a reference of exact values.

Author: @Rick-Anderson Do you know where this was supposed to point to? Things have moved since this link as added and I'm not sure what it should change to. Alternatively, we could drop it altogether...

Reviewer 2: I'd vote to drop it - the code doesn't have any more details than what is listed here.
